extension/src/providers/gotoProvider.ts

```diff
@@ -0,0 +1,145 @@
+				const varHeader = possibleDefun.getNthKeyAtom(symbolType.toUpperCase() === 'LAMBDA' ? 1 : 2);
+				if (innerContainer.line === varHeader.line && innerContainer.column === varHeader.column){
+					isFunction = false; // course correct for the first argument in a variable header being interpreted as a function call
+				}
+			}
+
+
+			// This condition determines the likelyhood of this being a function definition request or (else) a variable definition request
+			if (isFunction) {
+				let possible = this.findDefunMatches(selected, AutoLispExt.Documents.OpenedDocuments.concat(AutoLispExt.Documents.ProjectDocuments));
+				if (possible.length === 0) {
+					possible = this.findDefunMatches(selected, AutoLispExt.Documents.WorkspaceDocuments);
+				}
+				possible.forEach(item => {
+					locations.push(item);
+				});
+			} else if (parentContainer) {
+				this.findFirstVariableMatch(rDoc, position, selected, parentContainer).forEach(item => {
+					locations.push(item);
+				});
+			}
+			return locations;
+		} catch (error) {
+			return;	// I don't believe this requires a localized error since VSCode has a default "no definition found" response
+		}
+	}
+
+	
+	private findFirstVariableMatch(doc: ReadonlyDocument, start: vscode.Position, searchFor: string, searchIn: Sexpression): vscode.Location[] {
+		const result: vscode.Location[] = [];
+		const ucName = searchFor.toUpperCase();
+		let context = searchIn.getSexpressionFromPos(start);
+		let flag = true;
+		do {
+			const parent = searchIn.getParentOfSexpression(context);
+			const atom = parent?.getNthKeyAtom(0);
+			if (atom && /^DEFUN$|^DEFUN-Q|^LAMBDA$/i.test(atom.symbol)) {
+				let headers = parent.getNthKeyAtom(1);
+				if (headers?.symbol.toUpperCase() === ucName){ // adds defun names to possible result. Especially useful for quoted function names.
+					result.push(new vscode.Location(vscode.Uri.file(doc.fileName), new vscode.Position(headers.line, headers.column)));
+				}
+				if (!(headers instanceof Sexpression)){
+					headers = parent.getNthKeyAtom(2);
+				}
+				if (headers instanceof Sexpression){
+					const found = headers.atoms.find(p => p.symbol.toUpperCase() === ucName);
+					if (found) {
+						result.push(new vscode.Location(vscode.Uri.file(doc.fileName), new vscode.Position(found.line, found.column)));
+					}
+				}
+			} else if (atom && /^FOREACH$|^VLAX-FOR$/i.test(atom.symbol)) {
+				const tmpVar = parent.getNthKeyAtom(1);
+				if (!(tmpVar instanceof Sexpression) && tmpVar.symbol.toUpperCase() === ucName){
+					result.push(new vscode.Location(vscode.Uri.file(doc.fileName), new vscode.Position(tmpVar.line, tmpVar.column)));
+				}
+			}
+			if (!parent || result.length > 0){
+				flag = false;
+			} else {
+				context = parent;
+			}
+		} while (flag);
+		// If we still haven't found anything check the 1st occurrence of setq's. This will find globals setqs and nested ones possibly inside other defuns
```

Comment:
it is best to wrap to a small function

Reply:
I had started playing with the renameProvider and did begin breaking some stuff out that could be recycled. Never made it into this chunk of code, but I will look into it.